docs/thinking_in_jax.rst

```diff
@@ -0,0 +1,484 @@
+but are agnostic to the values. This recorded sequence of computations can then be efficiently
+applied within XLA to new inputs with the same shape and dtype, without having to re-execute the
+Python code.
+
+When we call the compiled fuction again on matching inputs, no re-compilation is required:
+
+.. ipython::
+
+   In [1]: x2 = np.random.randn(3, 4)
+
+   In [2]: y2 = np.random.randn(4)
+
+   @doctest
+   In [3]: f(x2, y2)
+   Out[3]: DeviceArray([1.4344584, 4.3004413, 7.9897013], dtype=float32)
+
+The extracted sequence of operations is encoded in a JAX expression, or *jaxpr* for short.
+You can view the jaxpr using the :func:`jax.make_jaxpr` transformation:
+
+.. ipython::
+
+   In [1]: from jax import make_jaxpr
+
+   In [2]: def f(x, y):
+      ...:     return jnp.dot(x + 1, y + 1)
+
+   @doctest
+   In [3]: make_jaxpr(f)(x, y)
+   { lambda  ; a b.
+     let c = add a 1.0
+         d = add b 1.0
+         e = dot_general[ dimension_numbers=(((1,), (0,)), ((), ()))
+                          precision=None ] c d
+     in (e,) }
+
+Note one consequence of this: because JIT compilation is done *without* information on the content
+of the array, control flow statements in the function cannot depend on traced values.
+For example, this fails:
+
+.. ipython::
+
+   In [1]: @jit
+      ...: def f(x, neg):
+      ...:   return -x if neg else x
+
+   @verbatim
+   In [2]: f(1, True)
+   ---------------------------------------------------------------------------
+   ConcretizationTypeError                   Traceback (most recent call last)
+   <ipython-input> in <module>()
+   ----> 1 f(1, True)
+   ...
+   ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected.
+
+If there are variables that you would not like to be traced,
+they can be marked as static for the purposes of JIT compilation:
+
+.. ipython::
+
+   In [1]: from functools import partial
+   
+   @verbatim
+   In [2]: @partial(jit, static_argnums=(1,))
+      ...: def f(x, neg):
+      ...:   return -x if neg else x
+
+   @suppress
+   In [3]: f = jit(lambda x, neg: -x if neg else x, static_argnums=(1,))  # IPython fails to parse the @partial() above.
+   
+   @doctest
+   In [3]: f(1, True)
+   Out[3]: DeviceArray(-1, dtype=int32)
+
+Note that calling a JIT-compiled function with a different static argument results in re-compilation,
+so the function still works as expected:
+
+.. ipython::
+   
+   @doctest
+   In [4]: f(1, False)
+   Out[4]: DeviceArray(1, dtype=int32)
+
+Understanding which values and operations will be static and which will be traced
+is a key part of using :func:`jax.jit` effectively.
+
+Static vs Traced Operations
+---------------------------
+
+**Key Concepts:**
+
+- Just as values can be either static or traced, operations can be static or traced.
+
+- Static operations are evaluated at compile-time in Python; traced operations are compiled
+  & evaluated at run-time in XLA.
```

Comment:
Nit:

```suggestion
  and evaluated at run-time in XLA.
```

(`&` -> `and` like in `JIT mechanics: tracing and static variables`)